objects/screen.c

```diff
@@ -297,18 +334,314 @@ screen_deduplicate(lua_State *L, screen_array_t *screens)
     }
 }
 
+/** Keep track of the screen viewport(s) independently from the screen objects.
+ *
+ * A viewport is a collection of `outputs` objects and their associated
+ * metadata. This structure is copied into Lua and then further extended from
+ * there. The `id` field allows to differentiate between viewports that share
+ * the same position and dimensions without having to rely on userdata pointer
+ * comparison.
+ *
+ * Screen objects are widely used by the public API and imply a very "visible"
+ * concept. A viewport is a subset of what the concerns the "screen" class
+ * previously handled. It is meant to be used by some low level Lua logic to
+ * create screens from Lua rather than from C. This is required to increase the
+ * flexibility of multi-screen setup or when screens are connected and
+ * disconnected often.
+ *
+ * Design rationals:
+ *
+ * * The structure is not directly shared with Lua to avoid having to use the
+ *   slow "miss_handler" and unsafe "valid" systems used by other CAPI objects.
+ * * The `viewport_t` implements a linked-list because its main purpose is to
+ *   offers a deduplication algorithm. Random access is never required.
+ * * Everything that can be done in Lua is done in Lua.
+ * * Since the legacy and "new" way to initialize screens share a lot of steps,
+ *   the C code is bent to share as much code as possible. This will reduce the
+ *   "dead code" and improve code coverage by the tests.
+ *
+ */
+typedef struct viewport_t
```

Comment:
Shouldn't this be declared/defined in `objects/screen.h`?

Reply:
I don't think it should, it is an implementation detail. It isn't used anywhere else in the code. It is an equivalent of `private` in Java/C++, also called "pimpl".

Reply:
Yes I see now that I dig into the code more. Was confused since I usually don't use typedefs and forward declare opaque structs in my headers. Was also confused since the implementation details do seem exposed to Lua, but not to the rest of the C code (not that those are related).